file_formats/cmt_format.ml

```diff
@@ -109,43 +111,75 @@ let output_cmt oc cmt =
   output_string oc Config.cmt_magic_number;
   output_value oc (cmt : cmt_infos)
 
-let read filename =
+let cmt_error filename = function
+  | Magic_number.Parse_error err ->
+     Not_a_typedtree (filename, err)
+  | Magic_number.Unexpected_error err ->
+     Unexpected_typedtree (filename, err)
+
+let cmi_error filename = function
+  | Magic_number.Parse_error err ->
+     Cmi_format.Not_an_interface (filename, err)
+  | Magic_number.Unexpected_error err ->
+     Cmi_format.Unexpected_interface (filename, err)
+
+let input_cmt_result filename ic : _ result =
+  match input_cmt ic with
+    | value -> Ok value
+    | exception _ -> Error (Corrupted_typedtree filename)
+
+let input_cmi_result filename ic : _ result =
+  match input_cmi ic with
+    | value -> Ok value
+    | exception _ -> Error (Cmi_format.Corrupted_interface filename)
+
+let read_results filename : _ result * _ result =
 (*  Printf.fprintf stderr "Cmt_format.read %s\n%!" filename; *)
   let ic = open_in_bin filename in
-  Misc.try_finally
-    ~always:(fun () -> close_in ic)
+  Fun.protect
+    ~finally:(fun () -> close_in ic)
     (fun () ->
-       let magic_number = read_magic_number ic in
-       let cmi, cmt =
-         if magic_number = Config.cmt_magic_number then
-           None, Some (input_cmt ic)
-         else if magic_number = Config.cmi_magic_number then
-           let cmi = Cmi_format.input_cmi ic in
-           let cmt = try
-               let magic_number = read_magic_number ic in
-               if magic_number = Config.cmt_magic_number then
-                 let cmt = input_cmt ic in
-                 Some cmt
-               else None
-             with _ -> None
-           in
-           Some cmi, cmt
-         else
-           raise(Cmi_format.Error(Cmi_format.Not_an_interface filename))
-       in
-       cmi, cmt
+      let open Magic_number in
+      match read_current_info ~expected_kind:None ic with
+        | Result.Error err ->
+           (Result.Error (cmi_error filename err),
+            Result.Error (cmt_error filename err))
+        | Result.Ok { kind; version = _ } ->
+      match kind with
```

Comment:
As I said before, I don't think you should use this indentation style in the compiler, given it's not a commonly used style, and could be misleading. Also, in this case, you can just inline the match if you want to avoid the nesting so much.

Reply:
Well I'm personally happy with this style. The strength of the counter-argument required to disregard a review comment could be proportionate to the importance of the change being discussed.

Reply:
Hang on, I don't think it's reasonable to dismiss this comment out of hand.
Part of the point of code review is ensuring that the code is easy for others to read, not just the author.

My personal view is that the compiler source isn't the place to experiment with notably different or unusual indentation styles, until such styles are reasonably widely adopted elsewhere. 

Specifically regarding this style, apart from the lack of indentation that could be confusing on the second match, using five spaces also seems rather odd.

Reply:
> My personal view is that the compiler source isn't the place to experiment with notably different or unusual indentation styles, until such styles are reasonably widely adopted elsewhere.

I suppose that this style is used elsewhere, given that it exists as an option in `ocamlformat` (`--nested-match=align`). (Also, I don't find your view particularly convincing; but I don't mind you using it in the code you write, including in the compiler.)

Re. number of spaces, I guess I used whatever Tuareg used by default, but I'm happy to move to whatever you prefer (4?) as a form of compromise.

Reply:
@gasche your indentation style is inconsistent: under `match` you indent the vertical bars by 2 spaces (yucky, especially when your continuation match gets dedented from its bar) but under `function` you use 0 or 2 spaces.

Reply:
I'm amused by the comment you decided to use to revive this PR. I apologize for leaving it dormant for so long, I know about it, but it is also super-boring to work on so I am going very slowly. Apologies.